File: qr-api/internal/middleware/error_handler.go
package middleware

import (
	"errors"
	"log"
	"net/http"

	"github.com/gofiber/fiber/v2"

	"github.com/gianfrancocespedes/fact-qr/qr-api/internal/apperror"
)

// errorBody es la forma que toda respuesta de error devuelve al cliente.
type errorBody struct {
	Error errorDetail `json:"error"`
}

type errorDetail struct {
	Code    apperror.Code    `json:"code"`
	Details apperror.Details `json:"details,omitempty"`
}

// statusByCode asigna el código HTTP a cada error de dominio.
var statusByCode = map[apperror.Code]int{
	apperror.CodeInvalidNumber:           http.StatusBadRequest,
	apperror.CodeEmptyMatrix:             http.StatusBadRequest,
	apperror.CodeInconsistentRowLength:   http.StatusBadRequest,
	apperror.CodeMatrixTooLarge:          http.StatusRequestEntityTooLarge,
	apperror.CodeInvalidRotations:        http.StatusBadRequest,
	apperror.CodeInvalidDirection:        http.StatusBadRequest,
	apperror.CodeInvalidPayload:          http.StatusBadRequest,
	apperror.CodeUnauthorized:            http.StatusUnauthorized,
	apperror.CodeTokenExpired:            http.StatusUnauthorized,
	apperror.CodeStatsServiceUnavailable: http.StatusServiceUnavailable,
	apperror.CodeInternal:                http.StatusInternalServerError,
}

// ErrorHandler es el punto único de serialización de errores de la API.
func ErrorHandler(c *fiber.Ctx, err error) error {
	var domainError *apperror.Error
	if errors.As(err, &domainError) {
		status, known := statusByCode[domainError.Code]
		if !known {
			status = http.StatusInternalServerError
		}
		return c.Status(status).JSON(errorBody{
			Error: errorDetail{Code: domainError.Code, Details: domainError.Details},
		})
	}

	// Fiber genera este error para rutas inexistentes y otros fallos de enrutamiento.
	var fiberError *fiber.Error
	if errors.As(err, &fiberError) {
		return c.Status(fiberError.Code).JSON(errorBody{
			Error: errorDetail{
				Code:    apperror.CodeInvalidPayload,
				Details: apperror.Details{"path": c.Path()},
			},
		})
	}

	// Cualquier otra cosa es un fallo no previsto.
	log.Printf("[qr-api] error no controlado: %v", err)
	return c.Status(http.StatusInternalServerError).JSON(errorBody{
		Error: errorDetail{Code: apperror.CodeInternal},
	})
}

File: stats-api/src/middleware/error-handler.js
import { AppError, ErrorCode } from '../errors/codes.js';

/**
 * Punto único de serialización de errores.
 *
 * Centralizarlo garantiza que toda respuesta de error tenga la misma forma
 * `{ error: { code, details } }`
 */
export function errorHandler(error, request, response, next) {
    if (error instanceof AppError) {
        response.status(error.status).json(error.toJSON());
        return;
    }

    // Un JSON malformado lo rechaza el parser de Express antes de llegar al controlador.
    if (error instanceof SyntaxError && 'body' in error) {
        response.status(400).json({ error: { code: ErrorCode.INVALID_PAYLOAD, details: { reason: 'malformed_json' } } });
        return;
    }

    // Cualquier otro error es un fallo no previsto.
    console.error('[stats-api] error no controlado:', error);
    response.status(500).json({ error: { code: ErrorCode.INTERNAL } });
}

/** Responde 404 con el mismo formato de error que el resto de la API. */
export function notFoundHandler(request, response) {
    response.status(404).json({
        error: { code: ErrorCode.INVALID_PAYLOAD, details: { path: request.path } },
    });
}
